README

General

  This is a Perl client for the network euchre server called euchred.

  This package is currently in alpha testing, so expect things to break.
  However, with some luck it can be used to play a complete game of
  euchre.

  The package is copyrighted by me, Denis McLaughlin, and licensed under
  the GNU Public License.  God bless Richard Stallman.


Denis McLaughlin
mclaughlin.denis@gmail.com

---------------------------------------------------------------------------

Requirements

    The client will need a working Perl 5 install on a Unix system.  The
    client depends on various terminal escape codes (screen clearing and
    text highlighting) which have been tested on the linux console as
    well as an xterm.  Other terminal types may need some tweaking.

---------------------------------------------------------------------------

Making It

  The client is written in Perl, and so doesn't need to be compiled.  You'll
  find the eucherl client executable in the src directory.

---------------------------------------------------------------------------

Installing It

  We're still in alpha testing, so there is no installer yet.  The
  client assumes it will be run from its src/ directory.

---------------------------------------------------------------------------

Running It

  To run it:
    - cd src
    - ./eucherl

  Eucherl is single player client, so to actually have a game of euchre
  you'll either have to run four clients and try not to cheat, or find
  three friends who also have the client.

  If you're running the clients and server on the same system, the
  client's default server setting is fine.  Otherwise reset the server
  by running ./eucherl -s <servername>, or using the 's' command in the
  client to reset the server name.

  Once you've started the client, there is a state-sensitive help menu.
  In brief, 'c' to connect, and once four players have joined (you'll
  see them in the display), 's' to start the game.  Up and down arrows
  scroll the chat messages forward and backward (space to zip to the
  bottom), left and right arrow highlight options (when available),
  return selects the current highlighted option.

---------------------------------------------------------------------------

Implementation Details

  See ./Implementation.

---------------------------------------------------------------------------

Bugs

  Bugs or other questions can be sent to mclaughlin.denis@gmail.com.

---------------------------------------------------------------------------


src/Changes

Perl Euchre Client Changes File

---------------------------------------------------------------------------

version 0.1

  May 10, 1999
  Denis McLaughlin
   - started perl euchre client
   - added Usage() function
   - added switch processing (-h, -p:, -s:)

  May 11, 1999
  Denis McLaughlin
   - added socket connection routine
   - added message reader

  May 13, 1999
  Denis McLaughlin
   - reorganized code
   - the main loop now accepts uncooked input, and runs a tight
     select() loop on both S and STDIN, although it is only processing
     STDIN commands

  May 15, 1999
  Denis McLaughlin
   - redid raw input, first using the Term::ReadKey package, then using
     just the POSIX termios stuff; the ReadKey stuff is tidier, but the
     termios code will be there by default, which is a key feature
   - finished the screen code: some hairiness, but it looks cool
   - added some more global variables for cards and players
   - added some message routines, to manage the list of displayed messages
   - I'll have to mung around to get key presses to work, be nice to do
     it generically
   - unbuffered stdout, so the last line will print

  May 16, 1999
  Denis McLaughlin
   - message formatting is completed: AddMessage() and watch the magic
   - woohoo, message scrolling works!
   - redirected all log messages through AddMessage
   - next: add server loss code 

  May 18, 1999
  Denis McLaughlin
   - Prepared initial code to send JOIN request and protocol version

  May 19, 1999
  Denis McLaughlin
   - reorganized code
   - sending JOIN request
   - problems receiving JOIN acknowledge?

  May 21, 1999
  Denis McLaughlin
   - removed ConfigChat routine with a require "messages.pl" call, to
     avoid potential conflicts with sysread() and select()
   - fixed select() problem, now receiving events, but looping?

  May 27, 1999
  Denis McLaughlin
   - tweaked the byte read section: I'm going to have to be aware of
     byte ordering issues to keep things portable
   - client receives JOINACCEPT, not processing yet
   - added perltags script, and a makefile for it

  May 31, 1999
  Denis McLaughlin
   - added a server die routine
   - added a proper JOINACCEPT routine
   - added a \r to the last line of the message display
   - woo hoo, network protocol works!

  June 1, 1999
  Denis McLaughlin
   - checked that multiple clients on a single server work
   - cleaned up screen
   - added a screen prompting routine
   - added a prompt to allow port number to be changed
   - up arrows broken?

  June 5, 1999
  Denis McLaughlin
   - all bytes being sent are munged for byte order
   - client can now understand JOINDENYs
   - client can now understand SERVERQUITs
   - client now sends CLIENTQUIT on exit

  June 6, 1999
  Denis McLaughlin
   - we now properly service DECLINEs, although when the server closes
     socket, it looks like a server disconnect to us
   - renamed OpenServer() to Connect(), and made a Disconnect(): on server
     quits or declines or join refusals, the client does not exit
   - we now clean $cmd between bad commands
   - arrow keys work again
   - name and server setting is now supported
   - messaging clean up
   - client can handle the KICK message
   - woo hoo, chat window shows ^^^^ when the screen is scroll backed, just
     like quake!

  June 12, 1999
  Denis McLaughlin
   - added an ID message, sent after connect, and after name change
   - client can now send a chat message
   - client can now receive a chat message
   - increasing hairiness, the prompt code now allows backspacing

  June 15, 1999
  Denis McLaughlin
   - all protocol strings will no be immediately preceded by their length,
     and may not be null terminated
   - rewrite ID message in light of this
   - oops, my mistake, it already conformed

  June 27, 1999
  Denis McLaughlin
   - reformatted the JOIN message to include the user's name
   - fixed timing problem with SendJoin and SendID
   - added command interlocking, preventing multiple connects, disconnects
     prior to connections, etc

  June 29, 1999
  Denis McLaughlin
   - changed all messages to be preceded by total message length
   - changed order from <gh> <ph> <msglen> to <msglen> <gh> <ph>
   - fixed bug in ID message

  July 6, 1999
  Denis McLaughlin
   - switched messages to have the msglen first rather than second, makes
     it easier for protocol checking
   - fixed bug with SERVERQUIT
   - hmmm, problem with server KICK messages?

  July 27, 1999
  Denis McLaughlin
   - Added $tail
   - Added SO_REUSEADDR to socket stuff

  July 28, 1999
  Denis McLaughlin
   - Added tails to outgoing messages
   - Added tail expectation on incoming messages: nothing happens if it's
     not there, but a message is posted

  July 29, 1999
  Denis McLaughlin
   - began adding state parse routine
   - redid player structure
   - cleaned up some keyboard stuff: $raw appears to be magic, and I
     don't care to investigate why
   - replaced the ugly sysread()s with ReadInt(), ReadString() and ReadShort()

  July 31, 1999
  Denis McLaughlin
   - added ReadBoolean()
   - redid screen to dynamically track partner and opponents
   - completed player and game status reading
   - added ^L as a valid command
   - still problem with multiple players and updates


src/game.pl

# Eucherl : Perl Euchre Client
#
# (C) Copyright 1999 Denis McLaughlin


# Right, let's do this puppy
sub PlayGame
{
	while (1)
	{
		UpdateScreen();
		ShowChat();

		$rout=Select();

		if (vec($rout,fileno(STDIN),1))
		{	ServiceSTDIN();
		}

		if ($S && vec($rout,fileno($S),1))
		{	ServiceS();
		}
	}
}


###########################################################################
# Deals with an unexpected server death

sub ServerDie
{
	AddChat("Client: Uh-oh, the server connection has been lost!\n");

	Disconnect();
}


###########################################################################
# This is the SetTeams() routine: it sets the teams, relying on the server
# always assigning teams in alternating order in the player structure.  We
# step through the player structure until we find ourself, then assign $foe1,
# $partner, and $foe2 sequentially after it.

sub SetTeams
{	local($pnum);

	$game{numplayers}=0;
	$me=""; $foe1=""; $partner=""; $foe2="";

	for ($pnum=0; $pnum<4; $pnum++)
	{
		if ($players{$pnum}{state} eq "joined")
		{	if ($players{$pnum}{ph} == $ph)
			{	$me=$pnum;
			}
			$game{numplayers}++;
		}
	}

	$foe1=($me+1)%4 if ($players{($me+1)%4}{state} eq "joined");
	$partner=($me+2)%4 if ($players{($me+2)%4}{state} eq "joined");
	$foe2=($me+3)%4 if ($players{($me+3)%4}{state} eq "joined");
}


###########################################################################
1;


src/socket.pl

# Eucherl : Perl Euchre Client
#
# (C) Copyright 1999 Denis McLaughlin

use Socket;


###########################################################################
# Routine to open and connect a socket to a remote server.
# This routine takes no arguments: it assumes that $port and $server (both
# global variables) are defined prior to entry.  The socket is passed back
# through $S, a global file descriptor.

sub Connect
{
	AddChat("Client: Attempting to connect to $server:$port.");

	$proto = getprotobyname('tcp');
	socket(S, PF_INET, SOCK_STREAM, $proto) ||
		Exit("Client: Error in socket(): $!");
	$sin = sockaddr_in($port,inet_aton($server));

	setsockopt(S,SOL_SOCKET,SO_REUSEADDR,1) ||
		Exit("Client: Error setting SO_REUSEADDR: $!\n");

	if (connect(S,$sin))
	{	$old=select(S); $|=1; select($old);
		$gstatus="connected";
		$S=S;
	}
	else
	{	AddChat("Client: Error in connect(): $!");
	}
}


###########################################################################
# Short routine to clean data structures following a server disconnect

sub Disconnect
{
	#AddChat("Client: Disconnecting from $server:$port.");
	AddChat("Client: Disconnecting from server.");

	CloseServer();

	ClearPlayer($foe1) if ("$foe1" ne "");
	ClearPlayer($foe2) if ("$foe2" ne "");
	ClearPlayer($partner) if ("$partner" ne "");
	ClearPlayer($me) if ("$me" ne "");

	ClearGame();

	$gstatus="unconnected";
}


###########################################################################
# Waits for an incoming message: returns when there is data to be read
# on STDIN or $S: returns $rout, the bit vector

sub Select
{
	local($rin,$rout);

	if ($S)
	{	undef $rin;
		vec($rin,fileno(STDIN),1)=1;
		vec($rin,fileno($S),1)=1;
		#AddChat("Setting STDIN and S: rin is ".unpack("B*",$rin)."\n");
	}
	else
	{	undef $rin;
		vec($rin,fileno(STDIN),1)=1;
		#AddChat("Setting STDIN: rin is ".unpack("B*",$rin)."\n");
	}

	raw();
	($nfound,$timeleft)=select($rout=$rin, undef, undef, undef);
	Exit("Client: select() error: $!\n") if ( $nfound < 0 );

	return($rout);
}


###########################################################################
# Shuts down the server socket politely

sub CloseServer
{
	if ($S)
	{	SendQuit("Sorry, it's been fun");
		close($S);
		$S=undef;
	}
}


###########################################################################
# Routine which takes a data buffer and sends it to the server socket

sub SendMessage
{
	local($data)=$_[0];

	print $S $data;
}


###########################################################################
# Short routine to create file descriptor vectors (read: file descriptor
# sets) for the select() routine.

sub fdset
{
	local(@fhlist) = split(' ',$_[0]);
	local($bits)=0;
	for (@fhlist)
	{
		vec($bits,fileno($_),1) = 1;
	}
	$bits;
}


###########################################################################
# Short routine to check if a certain file descriptor is set

sub fdisset
{
	local(@fhlist) = split(' ',$_[0]);
	local($bits);
	for (@fhlist)
	{
		vec($bits,fileno($_),1) = 1;
	}
	$bits;
}


###########################################################################
# Short routine to read a four byte integer off the server socket

sub ReadInt
{	local($tmp);

	$len=sysread($S,$tmp,4);

	# If the server sends us a message and then severs our link, sysread()
	# can sometimes return a system error (ENOLINK) prior to delivering the
	# data.  To work around this, we reread if we get a system error.
	$len=sysread($S,$tmp,4) if (!defined($len));

	return("") if ($len == 0);
    $tmp=unpack("N",$tmp);

	return($tmp);
}


###########################################################################
# Short routine to read a two byte short off the server socket

sub ReadShort
{	local($tmp);

	$len=sysread($S,$tmp,2);
	return("") if ($len == 0);
    $tmp=unpack("n",$tmp);

	return($tmp);
}


###########################################################################
# Short routine to read a string off the server socket

sub ReadString
{	local($tmp);
	local($len);

	sysread($S,$len,4);
    $len=unpack("N",$len);

	sysread($S,$tmp,$len);

	return($tmp);
}


###########################################################################
# Short routine to read a four byte integer, pretend it's a boolean and
# return it

sub ReadBoolean
{	local($tmp);

	$len=sysread($S,$tmp,4);
	return("") if ($len == 0);
    $tmp=unpack("N",$tmp);

	return($tmp);
}


###########################################################################
# Short routine to read a card from the socket: a card consists of two
# four-byte integers, the first representing the value (2 through 14),
# and the other representing the suit (0 through 3, in alphabetic order

sub ReadCard
{	local($tmp,$value,$suit);

	$value=ReadInt();
	$suit=ReadInt();

	$tmp="";
	$tmp = "$value" if ($value < 11);
	$tmp = "J" if ($value == 11);
	$tmp = "Q" if ($value == 12);
	$tmp = "K" if ($value == 13);
	$tmp = "A" if ($value == 14);

	$tmp .= "c" if ($suit == 0);
	$tmp .= "d" if ($suit == 1);
	$tmp .= "h" if ($suit == 2);
	$tmp .= "s" if ($suit == 3);
	
	return($tmp);
}


###########################################################################
# Short routine to read a suit from the socket: a suit consists of a
# four-byte integer, representing the suit (0 through 3, in alphabetic
# order

sub ReadSuit
{	local($tmp,$suit);

	$suit=ReadInt();

	$tmp .= "c" if ($suit == 0);
	$tmp .= "d" if ($suit == 1);
	$tmp .= "h" if ($suit == 2);
	$tmp .= "s" if ($suit == 3);
	
	return($tmp);
}


###########################################################################
# A short routine used to eat all pending data on server socket: only intended
# for debugging use

sub ClearS
{
	sysread($S,$tmp,10000);
}


###########################################################################
1;


src/screen.pl

# Eucherl : Perl Euchre Client
#
# (C) Copyright 1999 Denis McLaughlin


###########################################################################
# Routine to display current screen

sub UpdateScreen
{	local($pad,$gamestatus);

	$cname="Perl Euchre Client";
	$cnamel=length $cname;

	$srvprt="Server: $server  Port: $port";
	$srvprtl=length $srvprt;

	$statname="Status: $gstatus  Name: $name";
	$statnamel=length $statname;

	$gamestatus=GameStatus();
	$gamestatusl=length $gamestatus;

	$p0=PlayerName($me);
	$p1=PlayerName($foe1);
	$p2=PlayerName($partner);
	$p3=PlayerName($foe2);

	$p0l=length $p0; $p1l=length $p1; $p2l=length $p2; $p3l=length $p3;

	$p0string=PlayerString($me);
	$p1string=PlayerString($foe1);
	$p2string=PlayerString($partner);
	$p3string=PlayerString($foe2);

	$p0ls=length $p0string; $p1ls=length $p1string;
	$p2ls=length $p2string; $p3ls=length $p3string;

	# we don't set card values if $trickover so that we can see them
	if ( $trickover != 1 )
	{	$p0card=""; $p1card=""; $p2card=""; $p3card="";
		$p0card=Card($me);
		$p1card=Card($foe1);
		$p2card=Card($partner);
		$p3card=Card($foe2);
	}

	$p0lc=hilength($p0card); $p1lc=hilength($p1card);
	$p2lc=hilength($p2card); $p3lc=hilength($p3card);

	$options="";
	$options=OptionsString() if ( $gstatus eq "joined" );
	$optionsl=length $options;

	$trumpstring="";
	$trumpstring=TrumpString() if ( $game{ingame} );
	$trumpstringl=length $trumpstring;

	$choices=Choices();
	$choicesl=hilength($choices);

	# Display the score text if the game has started
	$score="";
	$score= "Score:  Us $game{score}{$players{$me}{team}}  Them $game{score}{$players{$foe1}{team}}" if ( $game{ingame});
	$scorel=length $score;

	$tricks="";
	$tricks= "Tricks: Us $game{tricks}{$players{$me}{team}}  Them $game{tricks}{$players{$foe1}{team}}" if ( $game{ingame});
	$tricksl=length $score;

	# set the help menu based on game state
    $help="q:Quit n:Name";
	$help.=" s:Server p:Port c:Connect" if ($gstatus eq "unconnected");
	$help.=" d:Disconnect" if ($gstatus ne "unconnected");
	$help.=" m:Msg" if ($gstatus eq "joined");
	$help.=" k:Kick" if ($players{$me}{creator});
	$help.=" o:Options" if ($players{$me}{creator} && !$game{ingame});
	$help.=" s:Start" if ($players{$me}{creator} && !$game{ingame}
		&& $game{numplayers}>3);
	$help.=" e:End" if ($players{$me}{creator});
	$pad = " "x(($msgcolumns-length($help))/2);
	$help = "$pad$help$pad\n";
	

	print "[H[J";

	# impending hirsuteness
	print "$cname"," "x($msgcolumns-($cnamel+$srvprtl)),"$srvprt\n";
	#print "$score     Status: $gstatus  Name: $name\n";
	print " "x($msgcolumns-$statnamel),"$statname\n";
	print "-"x$msgcolumns,"\n";
	print "$score"," "x($msgcolumns-($scorel+$optionsl)),"$options\n";
	print "$tricks"," "x($msgcolumns/2-($tricksl+int($p2l/2))),$p2," "x($msgcolumns-($tricksl+(int($msgcolumns/2)-($tricksl+int($p2l/2)))+$p2l+$trumpstringl)),"$trumpstring\n";
	print " "x(($msgcolumns-$p2ls)/2),$p2string,"\n";
	print "\n";
	print " "x(($msgcolumns-$p2lc)/2),$p2card,"\n";
	print "    ",$p1," "x((($msgcolumns/2)-5)-(4+$p1l+$p1lc)),$p1card," "x((($msgcolumns/2)+5)-(4+$p1l+((($msgcolumns/2)-5)-(4+$p1l+$p1lc))+$p1lc)),$p3card," "x($msgcolumns-(8+$p1l+$p1lc+$p3lc+$p3l+((($msgcolumns/2)-5)-(4+$p1l+$p1lc))+((($msgcolumns/2)+5)-(4+$p1l+((($msgcolumns/2)-5)-(4+$p1l+$p1lc))+$p1lc)))),$p3,"\n";
	print "    ",$p1string," "x((($msgcolumns/2)-1)-(4+$p1ls)),$p0card," "x($msgcolumns-(8+$p1ls+((($msgcolumns/2)-1)-(4+$p1ls))+$p0lc+$p3ls)),$p3string,"\n";
	#print "    ",$p1string," "x((($msgcolumns/2)-1)-(4+$p1ls)),$p0card," "x($msgcolumns-(8+$p1ls+((($msgcolumns/2)-1)-(4+$p1ls))+$p0lc+$p3ls)),$p3string,"\n";
	print "\n";
	print " "x(($msgcolumns-$p0l)/2),$p0,"\n";
	print " "x(($msgcolumns-$p0ls)/2),$p0string,"\n";
	print "$gamestatus\n";
	print "-"x$msgcolumns,"\n";
	print " "x(($msgcolumns-$choicesl)/2),$choices,"\n";
	print "-"x$msgcolumns,"\n";
    print "$help";
	print "-"x$msgcolumns,"\n";
};


###########################################################################
# A routine to append a message to the log: $lead is an internal use,
# to allow subsequent lines to be indented

sub AddChat
{
	local($msg)=$_[0];
	local($lead)=$_[1];
	local(@words,$word,$line)=(undef,undef,undef);

	# convert tabs to single spaces: sorry, but variable length characters
	# are just too awkward: note this is a client side limit, not a server
	# limit
	$msg =~ s/\t/ /g;

	# we remove all new lines, then add one at the end
	$msg =~ s/([^ \n])\n([^ \n])/\1 \2/g;
	$msg =~ s/\n//g;
	$msg .= "\n";

	# the simple case
	if ((length($msg)+length($lead)) < 79)
	{
		push @messages,"$lead$msg";
		UpdateScreen();
		ShowChat();
		return;
	}

	# okay, not so simple: split into words
	@words=split / /,$msg;

	# whoa, shit, no spaces at all! cut it in two, add the first, recurse
	if ($#words==0)
	{
		$word=substr $msg,0,78-length($lead);
		push @messages,"$lead$word\n";

		$word=substr $msg,78-length($lead);
		AddChat($word,"   ");
		return;
	}

	# pop words off the end of @words until @words is short enough
	while ((length("@words")+length($lead)) > 79)
	{
		$word=pop @words;
		$line="$word $line";
	}

	# then add @words (now short enough) and call recursively on $line
	push @messages,"$lead@words\n";
	AddChat("$line","   ");
	UpdateScreen();
	ShowChat();
}


###########################################################################
# Prints the appropriate set of messages from the end of @messages, taking
# into account the current value of $msgpos
#
# Done in three cases: $#messages < $msglines -> just print everything

sub ShowChat
{
	local($end,$line);

	$end=$#messages-$msgpos;
	$start=$end-($msglines-1); $start=0 if ($start < 0);
	
	for $i ($start..$end-1)
	{
		print $messages[$i];
	}

	if ($msgpos > 0)
	{	$line="^"x78;
	}
	else
	{	$line=$messages[$end];
	}

	$line =~ s/\n//;
	print $line;
	print "\r";
}


###########################################################################
# Used to show a prompt for further information.  The text prompt is
# expected to have been passed in as $_[0].  To keep things looking
# pretty, we null out any newlines in $_[0].  Because we use sysread()
# on STDIN elsewhere, we need to use a munged up little routine to
# read a line from STDIN here.
#
# GetString() returns with a single line of text, terminated by a
# newline.

sub GetString
{
	local($prompt)=$_[0];
	$prompt =~ s/\n//g;

	print "\n$prompt";

	#normal();
	raw();
	$string="";
	sysread STDIN,$char,1;
	while ($char ne "\n" && $char ne "\r")
	{	if (ord($char) == 127 || ord($char) == 8)
		{	$string=~s/.$//;
			UpdateScreen();
		}

		$string.=$char if (ord($char) => 32 && ord($char) < 127);
		UpdateScreen();
		print "\n$prompt$string";
		sysread STDIN,$char,1;
	}
	noecho();

	$string =~ s/\n//g;
	$string =~ s/$/\n/;

	return $string;
}


###########################################################################
# Used to set a new server

sub GetServer
{
	UpdateScreen();
	$newserver=GetString("Enter server name: ");
	$newserver =~ s/\n//g;

	if ($newserver =~ /^$/)
	{	AddChat("Client: Bad server specification: must enter name.\n");
		return;
	}

	$server=$newserver;
	AddChat("Client: New server is $newserver.\n");
	return;
}


###########################################################################
# Used to set a new port number

sub GetPort
{
	UpdateScreen();
	$newport=GetString("Enter port number: ");
	$newport =~ s/\n//g;

	if ($newport =~ /[^0-9]+/)
	{	AddChat("Client: Bad port specification: only numbers allowed.\n");
		return;
	}

	if ($newport !~ /[0-9]+/)
	{	AddChat("Client: Bad port specification: numbers required.\n");
		return;
	}

	$port=$newport;
	AddChat("Client: New port is $newport.\n");
	return;
}


###########################################################################
# Used to set a new name

sub GetName
{
	UpdateScreen();
	$newname=GetString("Enter name: ");
	$newname =~ s/\n//g;

	if ($newname =~ /^$/)
	{	AddChat("Client: Bad name: must enter name.\n");
		return;
	}

	$name=$newname;
	AddChat("Client: New name is $newname.\n");
	return;
}


###########################################################################
# Used to enter a message

sub GetChat
{
	$chat="";

	if ($gstatus ne "joined")
	{	AddChat("Client: Can't send message until connected.");
		return;
	}

	UpdateScreen();
	$chat=GetString("Enter message: ");
	$chat =~ s/\n//g;

	return;
}


###########################################################################
# Sets the values in a players screen string

sub PlayerString
{	local($pnum) = $_[0];
	local($tmp) = "";


	$tmp .= "creator," if ($players{$pnum}{creator});
	$tmp .= "dealer," if ($players{$pnum}{dealer});
	$tmp .= "maker," if ($players{$pnum}{maker});
	$tmp .= "leader," if ($players{$pnum}{leader});
	$tmp .= "order," if ($players{$pnum}{order});
	$tmp .= "alone," if ($players{$pnum}{alone});
	$tmp .= "defending," if ($players{$pnum}{defend});
	$tmp .= "play?," if ($players{$pnum}{playoffer});
	$tmp .= "order?," if ($players{$pnum}{orderoffer});
	$tmp .= "drop?," if ($players{$pnum}{dropoffer});
	$tmp .= "call?," if ($players{$pnum}{calloffer});
	$tmp .= "defend?," if ($players{$pnum}{defendoffer});

	$tmp =~ s/,$//;
	$tmp = "($tmp)" if ($tmp ne "");

	return($tmp);
}


###########################################################################
#  Sets the options string

sub OptionsString
{	local($options)="Options: ";

	$options.="a" if (  $game{aloneonorder} );
	$options.="-" if ( !$game{aloneonorder} );
	$options.="d" if (  $game{candefend} );
	$options.="-" if ( !$game{candefend} );
	$options.="s" if (  $game{screw} );
	$options.="-" if ( !$game{screw} );

	return($options);
}


###########################################################################
#  Sets the trump string

sub TrumpString
{	local($trumpstring)="";

	$trumpstring = "Clubs" if ( $game{trump} eq "c" );
	$trumpstring = "Diamonds" if ( $game{trump} eq "d" );
	$trumpstring = "Hearts" if ( $game{trump} eq "h" );
	$trumpstring = "Spades" if ( $game{trump} eq "s" );
	#$trumpstring = "<not set>" if ( $trumpstring eq "" );
	$trumpstring = "<not set>" if ( !$game{trumpset} );

	return("Trump: $trumpstring");
}


###########################################################################
# Prints a player structure 

sub PrintPlayer
{	local($pnum)=$_[0];

	AddChat("state is $players{$pnum}{state}");
	AddChat("ph is $players{$pnum}{ph}");
    AddChat("name is $players{$pnum}{name}");
    AddChat("clientname is $players{$pnum}{clientname}");
    AddChat("hw is $players{$pnum}{hardware}");
    AddChat("os is $players{$pnum}{os}");
    AddChat("comment is $players{$pnum}{comment}");
    AddChat("team is $players{$pnum}{team}");
    AddChat("numcards is $players{$pnum}{numcards}");
    AddChat("creator? $players{$pnum}{creator}");
    AddChat("ordered? $players{$pnum}{ordered}");
    AddChat("dealer? $players{$pnum}{dealer}");
    AddChat("alone? $players{$pnum}{alone}");
    AddChat("defending? $players{$pnum}{defend}");
    AddChat("leading? $players{$pnum}{leader}");
    AddChat("play? $players{$pnum}{playoffer}");
    AddChat("order? $players{$pnum}{orderoffer}");
    AddChat("call? $players{$pnum}{calloffer}");
    AddChat("defend? $players{$pnum}{defendoffer}");
    AddChat("cardinplay? $players{$pnum}{cardinplay}");
    AddChat("card $players{$pnum}{card}");
}


###########################################################################
# sets the value for a players name intelligently

sub PlayerName
{	local($pnum)=$_[0];
	local($tmp);

	$tmp = "$players{$pnum}{name} ($players{$pnum}{numcards})"
		if ($players{$pnum}{state} eq "joined");

	$tmp =~ s/ *\(\) *//;

	return($tmp);
}


###########################################################################
# sets the value for the game status line

sub GameStatus
{
	return("Waiting to connect...")
		if ( $gstatus eq "unconnected");

	return("Waiting for players...")
		if ( ! $game{ingame} && $game{numplayers} < 4);

	return("Waiting for creator to start...")
		if ( ! $game{ingame} && $game{numplayers} == 4);

	return("Offer to order trump...")
		if ( $game{hstate} == 1);

	return("Offer to call trump...")
		if ( $game{hstate} == 2);

	return("Offer to defend alone...")
		if ( $game{hstate} == 3);

	return("Playing...")
		if ( $game{hstate} == 4);

	return("Unknown game state");
}


###########################################################################
# Used to specify a player, probably to kick them off

sub GetPlayer
{	local($targetph)="";
	local($player)="";
	local($pnum);

	UpdateScreen();
	$player=GetString("Player to kick: ");
	$player =~ s/\n//g;

	for ($pnum=0; $pnum<4; $pnum++)
	{	if ($players{$pnum}{state} eq "joined")
		{	if ("$players{$pnum}{name}" eq "$player")
			{	$targetph=$players{$pnum}{ph};
			}
		}
	}

	if ($targetph eq "")
	{	AddChat("Client: Can't find player $player");
		return("");
	}

	return($targetph);
}


###########################################################################
# Gets options, returning them in an array

sub GetOptions
{	local($candefend,$aloneonorder,$screw);

	UpdateScreen();
	$aloneonorder=GetAnswer("Do players go alone after ordering their partner?",
		$game{aloneonorder});

	UpdateScreen();
	$candefend=GetAnswer("Can players defend alone?",$game{candefend});

	UpdateScreen();
	$screw=GetAnswer("Is screw the dealer in effect?",$game{screw});

	return($candefend,$aloneonorder,$screw);
}


###########################################################################
# Used to show a yes/no prompt.  The current value of the variable is passed
# as the second argument, and is used to compose the default, either [Y/n]
# or [y/N].  The only way to change the default value is to press the letter
# (case insensitive) corresponding to the opposite: all other keys will
# keep the current value.

sub GetAnswer
{
	local($prompt)=$_[0];
	local($curval)=$_[1];
	$prompt =~ s/\n//g;

	$prompt .= " [Y/n] " if ($curval);
	$prompt .= " [y/N] " if (!$curval);

	print "'n' to change the value, anything else to leave it unchanged\n"
		if ($curval);
	print "'y' to change the value, anything else to leave it unchanged\n"
		if (!$curval);
	print "\n$prompt";

	sysread STDIN,$char,1;
	$char =~ tr /A-Z/a-z/;

	return(0) if ($char eq "n" && $curval);
	return(1) if ($char eq "y" && !$curval);
	return(1) if ($curval);
	return(0) if (!$curval);
}


###########################################################################
# This routine computes the whether anything should be displayed for that
# players card-in-play string

sub Card
{	local($pnum)=$_[0];

	if ($game{hstate}==1)
	{	if ($players{$pnum}{dealer})
		{	return($game{hole});
		}
		else
		{	return("");
		}
	}

	if ($game{hstate}==4)
	{	return($players{$pnum}{card})
			if ( $players{$pnum}{cardinplay} == 1 && $trickover == 0);
	}
}


###########################################################################
# A routine to return a string of highlighted text

sub hi
{	return "$hion$_[0]$hioff";
}


###########################################################################
# A routine that returns the current choice string

sub Choices
{	local($cards);

	return "" if ( !$game{ingame});
	$cards=$players{$me}{cards};

	# continue, after trick
	return("Cards: ".$cards."    ".
		HiChoices(" ","Continue"))
		if ( $trickover == 1 );

	# order the hole card up if your partner is the dealer and screw is on
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Pass","Order alone"))
		if ($players{$me}{orderoffer} && $players{$partner}{dealer} &&
			$game{aloneonorder} );

	# order the hole card up
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Pass","Order","Order alone"))
		if ( $players{$me}{orderoffer} && !$players{$me}{dealer});

	# pick up the hole card
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Pass","Pick up","Pick up alone"))
		if ( $players{$me}{orderoffer} && $players{$me}{dealer});

	# having been ordered or having picked up, drop a card
	return("Drop Card: ".
		HiChoices(" ",split ' ',$cards))
		if ( $players{$me}{dropoffer});

	# call a trump if you're not the dealer
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Pass","Call","Call alone"))
		if ( $players{$me}{calloffer} && !$game{callstate}
			&& !$players{$me}{dealer} );

	# call a trump if you're a dealer and screw the dealer is set
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Call","Call alone"))
		if ( $players{$me}{calloffer} && !$game{callstate}
			&& $players{$me}{dealer} && $game{screw} );

	# call a trump if you're a dealer and screw the dealer is not set
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Fold hand","Call","Call alone"))
		if ( $players{$me}{calloffer} && !$game{callstate}
			&& $players{$me}{dealer} && !$game{screw} );

	# choose a suit to call
	return("Cards: ".$cards."    ".
		HiChoices(" : ",CallSuits()))
		if ( $players{$me}{calloffer} && $game{callstate});

	# defend or not defend
	return("Cards: ".$cards."    ".
		HiChoices(" : ","Pass","Defend alone"))
		if ( $players{$me}{defendoffer});

	# choose a card to play
	return("Play Card: ".
		HiChoices(" ",split ' ',$cards))
		if ( $players{$me}{playoffer});

	return ("Cards: $cards") if ( $game{ingame});
}


###########################################################################
# This routine returns a list of choices, appropriately highlighted.

sub HiChoices
{	local($i,$chc);
	local($result)="";
	local($separator)=shift(@_);

	$i=0;
	$choice=$#_ if ( $choice > $#_);

	for $chc (@_)
	{	if ($i == $choice)
		{	$result .= hi($chc);
			$choicetext=$chc;
		}
		else
		{	$result .= $chc;
		}

		$result .= $separator;
		$i++;
	}

	$result =~ s/[ :]*$//;
	return($result);
}


###########################################################################
# This routine returns the list of valid suits that can be called

sub CallSuits
{	local($suit,@result);

	for $suit ("c","d","h","s")
	{	push @result,SuitToText($suit) if ( $suit ne SuitOf($game{hole}) );
	}
	push @result,"Cancel";

	return(@result);
}


###########################################################################
# Changes the suit string to a name

sub SuitToText
{	return("Clubs") if ($_[0] eq "c");
	return("Diamonds") if ($_[0] eq "d");
	return("Hearts") if ($_[0] eq "h");
	return("Spades") if ($_[0] eq "s");
	return("Unknown Suit");
}


###########################################################################
# Changes the suit name to a suit type

sub TextToType
{	return(0) if ($_[0] eq "Clubs");
	return(1) if ($_[0] eq "Diamonds");
	return(2) if ($_[0] eq "Hearts");
	return(3) if ($_[0] eq "Spades");
	return(-1);
}


###########################################################################
# takes a card of the form "6h" and returns "h"

sub SuitOf
{	local($card)=$_[0];

	$card =~ s/^.*(.)$/\1/g;
	return($card);
}


###########################################################################
# takes a card of the form "6h" and returns the value and suit integers,
# appropriate to be sent in a network message

sub CardToType
{	local($card)=$_[0];

	($value=$card) =~ s/^(.*).$/\1/g;
	$value=11 if ($value eq "J");
	$value=12 if ($value eq "Q");
	$value=13 if ($value eq "K");
	$value=14 if ($value eq "A");

	($suit=$card) =~ s/^.*(.)$/\1/g;
	$suit=0 if ($suit eq "c");
	$suit=1 if ($suit eq "d");
	$suit=2 if ($suit eq "h");
	$suit=3 if ($suit eq "s");

	return($value,$suit);
}


###########################################################################
# This routine dynamically resizes the screen

sub Winch
{	
	$msglines=6;
	$msglines=$ENV{LINES}-19
		if ( defined($ENV{LINES}) );

	$msgcolumns=79;
	$msgcolumns=$ENV{COLUMNS}-1
		if ( defined($ENV{COLUMNS}) );
}


###########################################################################
# This routine returns the length of a string, discounting any hilighting
# characters

sub hilength
{	local($string)=$_[0];

	$string =~ s/\[K//;
	$string =~ s/\[7m//;
	$string =~ s/\[m//;
	$string =~ s/\[27m//;

	return(length($string));
}


###########################################################################
1;
